### PR TITLE
CV-21233 Migrating to GD

### PR DESCRIPTION
**Pull request by ikonic89**
Jira: https://chesscom.atlassian.net/browse/CV-21233

- Migrating to GD using [Intervention/image](https://github.com/Intervention/image)
- Code refactor, to make things simpler

Todo:
- [x] test generated image size & compare with old version
- [x] test & compare speed & number of disk writes
- [x] flipping not working correctly
- [x] docs
- [x] clean-up old classes

cc: 
@lackovic10 for the later review
@BiviaBen you worked with Ivan a lot on this, so if you want, take a peak

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -18,7 +18,9 @@
         "jms/serializer": "~0.12",
         "symfony/routing": "~2.1",
         "symfony/yaml": "~2.1",
-        "symfony/validator": "~2.1"
+        "symfony/validator": "~2.1",
+        "symfony/validator": "~2.1",
```
Comment (ikonic89) [inline composer.json]: Looks like it :)

**Comment by BiviaBen** [comment]
I'm working my way through this. Can we address the CI failures at the same time?

**Comment by ikonic89** [comment]
Yes, gonna work on it in an hour or two

File: src/DiagramGenerator/Board.php [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+
+        @mkdir($this->rootCacheDir.'/'.$this->cacheDirName, 0777);
+
+        $this->fen = Fen::createFromString($this->config->getFen());
+
+        if ($this->config->getFlip()) {
+            $this->fen->flip();
+        }
+
+        $this->image = $this->generateImage();
```
Comment (umpirsky) [inline src/DiagramGenerator/Board.php]: Why there, why not in `getImage()` when it's really needed?

File: src/DiagramGenerator/Image/Storage.php [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+    /**
+     * Fetches remove file, and stores it locally
+     *
+     * @param $remoteImageUrl
+     * @param $cachedFilePath
+     */
+    protected function cacheImage($remoteImageUrl, $cachedFilePath)
+    {
+        $cachedFilePathTmp = $cachedFilePath.uniqid('', true);
+        $ch = curl_init($remoteImageUrl);
```
Comment (umpirsky) [inline src/DiagramGenerator/Image/Storage.php]: Add `ext-curl` to `composer.json` requirements.

File: src/DiagramGenerator/Board.php (hunk 2 above)
Comment (ikonic89) [inline src/DiagramGenerator/Board.php]: Well, I consider `Board` class just a wrapper for this image, and instance of `Board` would be in invalid state if it doesn't have the image. 

Although, I do understand your point. I think I'm gonna change this.

Comment (umpirsky) [inline src/DiagramGenerator/Board.php]: From outside, all I care is to get image, I don't care at which point image will be created. That makes it lazy, not invalid. :)

**Comment by ikonic89** [comment]
Tests fixed, Sasa's comments addressed.

File: src/DiagramGenerator/Board.php [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+    public function __construct(Config $config, $rootCacheDir, $boardTextureUrl, $pieceThemeUrl)
+    {
+        $this->config = $config;
+        $this->rootCacheDir = $rootCacheDir;
+        $this->boardTextureUrl = $boardTextureUrl;
+        $this->pieceThemeUrl = $pieceThemeUrl;
+
+        $this->cacheDir = $this->rootCacheDir.'/'.$this->cacheDirName;
+
+        @mkdir($this->rootCacheDir.'/'.$this->cacheDirName, 0777);
```
Comment (BiviaBen) [inline src/DiagramGenerator/Board.php]: Is there a reason to concatenate the same thing twice?
```php
$this->cacheDir = $this->rootCacheDir.'/'.$this->cacheDirName;
@mkdir($this->cacheDir, 0777);
```
Is this the only place `$cacheDir` appears?

File: src/DiagramGenerator/Board.php [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+    }
+
+    /**
+     * Controls board image generating flow
+     *
+     * @return Image
+     */
+    protected function generateImage()
+    {
+        $storage = new Storage($this->rootCacheDir, $this->pieceThemeUrl, $this->boardTextureUrl);
```
Comment (BiviaBen) [inline src/DiagramGenerator/Board.php]: Here you pass `$this->rootCacheDir` and then in the Storage class you append a hard-coded "diagram_generator" string... to achieve the same thing as `$this->cacheDir` that was created during instantiation.

File: src/DiagramGenerator/Image/Image.php [hunk 6]
```diff
@@ -0,0 +1,292 @@
...
+use DiagramGenerator\Config;
+use DiagramGenerator\Config\Texture;
+use DiagramGenerator\Board;
+use DiagramGenerator\Fen;
+use DiagramGenerator\Generator;
+use Intervention\Image\Gd\Decoder;
+use Intervention\Image\Gd\Font;
+use Intervention\Image\Image as BaseImage;
+
+// TODO FLIPPING
```
Comment (BiviaBen) [inline src/DiagramGenerator/Image/Image.php]: Does this still need to be done?

Comment (ikonic89) [inline src/DiagramGenerator/Image/Image.php]: Nope, it's finished :)

File: src/DiagramGenerator/Image/Storage.php [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+     *
+     * @return Image
+     */
+    public function getBackgroundTextureImage(Config $config)
+    {
+        $boardCachedPath = $this->getCachedTextureFilePath($config);
+
+        try {
+            return ImageManagerStatic::make($boardCachedPath);
+        } catch (NotReadableException $exception) {
```
Comment (BiviaBen) [inline src/DiagramGenerator/Image/Storage.php]: Is this specific exception going to cover all the things this `catch` repairs? If the file is missing, corrupted, partial, or not actually an image?

**Comment by BiviaBen** [comment]
This is great. Only minor comments that are more like "refactoring" than what this PR is for. You get 🍰 !

File: src/DiagramGenerator/Image/Storage.php (hunk 7 above)
Comment (ikonic89) [inline src/DiagramGenerator/Image/Storage.php]: Yeah. Vendor basically calls `getimagesize()`, and if it fails, throws this exception. So, it covers all your cases.

**Pull request merged**